docs/RFCS/20170920_sql_consistency_checks.md

```diff
@@ -0,0 +1,428 @@
+
+The SQL checks, as they are composed of SQL, are fairly easy to express
+and create. For example, we can check a secondary index size using[1]:
+```sql
+# Get the index size
+SELECT count(*) FROM table@index
+# Get the table size
+SELECT count(*) FROM table
+# Then assert they are equal
+```
+
+For KV checks they will be executed in distSQL processors.
+
+TODO: High level of what happens when we fail a check? (Detailed version
+below)
+
+[1]: This isn't a fully correct way to check full secondary index
+consistency. There are problems that may impact the index size for a net
+zero result, causing false positives.
+
+# Detailed design
+
+Key parts of designing the checks are:
+- How are they invoked and executed
+- How will each check be implemented
+- What do we do when a check fails
+- How we can repair various failures
+
+## Invocation and execution
+
+As mentioned earlier, all of the checks are invokable by the CLI.
+
+The checks can also be run on a schedule. The frequency can be
+configured via a CLI setting (TBD) and with a command line parameter
+`--consistency-check-frequency`. For example,
+`--consistency-check-frequency=24H` or
+`--consistency-check-frequency=2D`.
+
+TODO: How will the SQL checks get incorporated if needed? What happens
+when we begin the FSCK call? I'm assuming we can rewrite the expression to make a
+custom entry point?
+
+To run KV checks we can execute our check logic within a distSQL
+processor[2].
+
+[2]: Implementation detail: we will either need to create the distSQL
+plans manually or create a way to express custom distSQL plans in SQL.
+The latter probably warrants more discussion, so won't be the likely
+option.
+
+## Implementations of checks
+
+SQL checks on the table schema can be implemented in an already existing
+`TableDescriptor.Validate`. The new capabilities are:
+- Check for invalid parent ID for interleaved tables
+- Check for invalid index definitions (wrong columns)
+- Check for invalid view descriptors: invalid column types; invalid SQL; invalid
+  number of columns.
+- Check for invalid column families
+
+The following checks are on the logical table, and so are possible
+through SQL checks[3].
+- `CHECK` constraints
+- `NOT NULL` constraints
+- `UNIQUE` constraints
+- `FOREIGN KEY` constraints
+- Secondary index entries exist
+- Dangling index references (i.e. missing row but index entry retained)
+
+The rest of the checks are on the physical table, and so require KV
+checks:
+- Check for indexes using [STORING][] have the correct data
+- Tables using [FAMILY][] have the data organized into families
+  correctly
+- Invalid data encodings
+- Non-canonical encodings (i.e. data round-trips)
+- [Composite encodings][] are valid
+- Key encoding correctly reflects the ordering of the decoded values
+
+[3]: An exception is secondary indexes, which may be possible to check
+on the SQL layer. The skepticism is that simply checking the size isn't
+suffice.
```

Comment:
yes simply checking the size is insufficient. You can run an UPDATE that updates the row and doesn't update the index. the index size will remain the same but be broken